agents/subtask_predictor.py

```diff
@@ -0,0 +1,311 @@
+        hidden = self.rnn(inp)
+        
+        # pass through fc to get logits
+        out = self.fc(hidden)
+
+        return out
+    
+    def init_hidden(self, batch_size):
+        # This method generates the first hidden state of zeros which we'll use in the forward pass
+        hidden = th.zeros(batch_size, Subtasks.NUM_SUBTASKS)
+        hidden[:,11]  = 1
+        return hidden
+    
+class SP_trainer():
+    def __init__(self, env, encoding_fn, dataset, args):
+        self.env = env
+        self.encode_state_fn = encoding_fn
+        self.remove_list = []
+        self.dataset = dataset
+        self.device = th.device('cuda' if th.cuda.is_available() else 'cpu')
+        self.num_players = 2
+        self.env = env
+        self.encode_state_fn = encoding_fn
+        self.args = args
+        visual_obs, agent_obss = self.encode_state_fn(self.env.mdp, self.env.state, self.args.horizon)
+        visual_obs_shape = visual_obs[0].shape
+        agent_obs_shape = agent_obss[0].shape
+        self.players = (
+            SubtaskPredictor(self.device, visual_obs_shape, agent_obs_shape),
+            SubtaskPredictor(self.device, visual_obs_shape, agent_obs_shape)
+        )
+        self.hidden = [self.players[0].init_hidden(args.batch_size), self.players[1].init_hidden(args.batch_size)]
+        #define optimizers and loss function
+        if dataset is not None:
+            self.optimizers = tuple([th.optim.Adam(player.parameters(), lr=args.lr) for player in self.players])
+            self.subtask_criterion = nn.CrossEntropyLoss(weight=th.tensor(dataset.get_subtask_weights(), dtype=th.float32, device=self.device),
+                                                             reduction='none')
+  
+    
+    def train_on_batch(self, batch):
+        """
+        Train BC agent on a batch of data
+        """
+        batch = {k: v.to(self.device) for k,v in batch.items()}
+        vo, ao, action, subtasks = batch['visual_obs'].float(), batch['agent_obs'].float(), \
+                                   batch['joint_action'].long(), batch['subtasks'].long()
+
+        curr_subtask, next_subtask = subtasks[:, 0], subtasks[:, 1]
+        metrics = {}
+        
+        
+        for i in range(self.num_players):
+            self.optimizers[i].zero_grad()
+            cs_i = F.one_hot(curr_subtask[:,i], num_classes=Subtasks.NUM_SUBTASKS)
+            sub_history = []
+            # create subtask history of past 10 subtasks
+            for idx, sample in enumerate(cs_i):
+                past_history = []
+                if idx>10:
+                    # append past 10 subtasks 
+                    for ids in range(idx-10,idx):
+                        past_history.append(cs_i[ids])
+                else:
+                    for ids in range(0,10):
+                        past_history.append( [0]*12)
+                        past_history[ids][11]=1
```

Comment:
Although the instances within the batch are no longer shuffled, there are still disconnects between trials and you don't have any logic to catch those instances.